File: PortTunnelLib/Java/PortTunnel.java
package PortTunnelLib;

import java.io.*;
import java.util.*;

public class PortTunnel {
    //-------------------------Private Attributes------------------------------
    private int mode;

    //--------------------Package Private Attributes---------------------------
    Connection Connection1;
    Connection Connection2;
    int ControlPort;
    int Timeout;

    boolean destIsSsl = false;
    boolean httpProtocol = false;
    int TicksBetweenBytes = 1;

    int Status;
    int ActiveConnections = 0;
    List<StreamLogListener> StreamLogListeners = new ArrayList<>();
    List<LogListener> LogListeners = new ArrayList<>();

    //---------------------------Public Methods--------------------------------
    /// <summary>
    /// Constructor for command line args
    /// </summary>
    /// <param name="args">Arguments for PortTunnel</param>
    public PortTunnel(String[] args)
    {
        Log.main = this;
        boolean configured = true;

        String tmp = null;
        Map<String, String> pairs = new HashMap<>();
        pairs.put("-m", null);
        pairs.put("-h1", null);
        pairs.put("-p1", null);
        pairs.put("-h2", null);
        pairs.put("-p2", null);
        pairs.put("-cp", null);
        pairs.put("-t", null);

        AddListener(new LogListener() {
            @Override
            public void NewLogOutput(String text) {
                System.out.print(text);
            }
        });

        if (args.length > 0)
        {
            if (args.length > 1)
            {
                for (String arg : args)
                {
                    if (arg.equals("-http"))
                    {
                        httpProtocol = true;
                    }
                    else if (arg.equals("-ssl"))
                    {
                        destIsSsl = true;
                    }
                    else if (arg.equals("-log"))
                    {
                        AddListener(new StreamLogListener() {
                            @Override
                            public void NewStreamLog(String text) {
                                String filename = text.split("\r")[0]
                                        .replace(' ', '-')
                                        .replace(":", "-")+ ".txt";

                                try {
                                    new File("Log/").mkdirs();

                                    BufferedWriter writer = new BufferedWriter(new FileWriter("Log/" + filename));
                                    writer.write(text);

                                    writer.close();
                                }catch (Exception e){
                                    e.printStackTrace();
                                }
                            }
                        });
                    }
                    else
                    {
                        if (tmp == null)
                            tmp = arg;
                        else
                        {
                            try
                            {
                                pairs.replace(tmp, arg);
                            }
                            catch (Exception e)
                            {
                                Log.Write(LogType.Warning, "Unknown argument: " + tmp);
                            }
                            tmp = null;
                        }
                    }
                }

                if (pairs.get("-m").toLowerCase().equals("sc"))
                {
                    if (pairs.get("-h1") != null && pairs.get("-p1") != null && pairs.get("-p2") != null)
                    {
                        try
                        {
                            Connection1 = new Connection(pairs.get("-h1"), Integer.parseInt(pairs.get("-p1")));
                            Connection2 = new Connection(Integer.parseInt(pairs.get("-p2")));
                        }
                        catch (Exception e)
                        {
                            Log.Write(LogType.Error, "Hostname or port not supported!");
                            configured = false;
                        }
                    }
                    else
                    {
                        Log.Write(LogType.Error, "Arguments not set for mode " + pairs.get("-m"));
                        configured = false;
                    }
                }
                else if (pairs.get("-m").toLowerCase().equals("cc"))
                {
                    if (pairs.get("-h1") != null && pairs.get("-p1") != null &&
                            pairs.get("-h2") != null && pairs.get("-p2") != null && pairs.get("-cp") != null)
                    {
                        try
                        {
                            Connection1 = new Connection(pairs.get("-h1"), Integer.parseInt(pairs.get("-p1")));
                            Connection2 = new Connection(pairs.get("-h2"), Integer.parseInt(pairs.get("-p2")));
                            ControlPort = Integer.parseInt(pairs.get("-cp"));
                        }
                        catch (Exception e)
                        {
                            Log.Write(LogType.Error, "Hostname or port not supported!");
                            configured = false;
                        }
                    }
                    else
                    {
                        Log.Write(LogType.Error, "Arguments not set for mode " + pairs.get("-m"));
                        configured = false;
                    }
                }
                else if (pairs.get("-m").toLowerCase().equals("ss"))
                {
                    if (pairs.get("-p1") != null && pairs.get("-p2") != null && pairs.get("-cp") != null)
                    {
                        try
                        {
                            Connection1 = new Connection(Integer.parseInt(pairs.get("-p1")));
                            Connection2 = new Connection(Integer.parseInt(pairs.get("-p2")));
                            ControlPort = Integer.parseInt(pairs.get("-cp"));
                        }
                        catch (Exception e)
                        {
                            Log.Write(LogType.Error, "Hostname or port not supported!");
                            configured = false;
                        }
                    }
                    else
                    {
                        Log.Write(LogType.Error, "Arguments not set for mode " + pairs.get("-m"));
                        configured = false;
                    }
                }
                else
                {
                    Log.Write(LogType.Error, "Unknown mode!");
                    configured = false;
                }
            }
            else
            {
                configured = false;
                if(args[0].equals("-h"))
                {
                    Log.Write(LogType.Info, "\r\n" + GetHelp());
                }
            }
        }
        else
        {
            Log.Write(LogType.Error, "No args found!");
            configured = false;
        }

        try
        {
            Timeout = Integer.parseInt(pairs.get("-t"));
        }
        catch (Exception e)
        {
            Timeout = 3000;
        }

        if (!configured)
        {
            Log.Write(LogType.Info, "Exiting now");
            ExitEnvironment();
        }
    }

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="connection1">Object</param>
    /// <param name="connection2">Object</param>
    /// <param name="timeout">Timeout for each tunnel</param>
    /// <param name="controlPort">PortTunnel communication port</param>
    public PortTunnel(Connection connection1, Connection connection2,
                      int timeout, int controlPort)
    {
        Log.main = this;

        Connection1 = connection1;
        Connection2 = connection2;
        ControlPort = controlPort;
        Timeout = timeout;
    }

    /// <summary>
    /// Starts PortTunnel
    /// </summary>
    /// <returns>false if config is not supported</returns>
    public boolean Start()
    {
        boolean supported = true;
        Status = 1;

        Thread Starter = null;

        if (Connection1.Type == ConnectionType.Client &&
                Connection2.Type == ConnectionType.Server)
        {
            mode = 0;
            Starter = new ServerClient(this);
        }
        else if (Connection1.Type == ConnectionType.Client &&
                Connection2.Type == ConnectionType.Client)
        {
            mode = 1;
            Starter = new ClientClient(this);
        }
        else if (Connection1.Type == ConnectionType.Server &&
                Connection2.Type == ConnectionType.Server)
        {
            mode = 2;
            Starter = new ServerServer(this);
        }
        else
        {
            supported = false;
            Log.Write(LogType.Error, "Config not supported");
        }

        ShowConfig();

        if (Starter != null)
            Starter.start();

        return supported;
    }

    /// <summary>
    /// Stops PortTunnel
    /// </summary>
    public void Stop()
    {
        Status = 0;
    }

    /// <summary>
    /// Exits the main process
    /// </summary>
    public static void ExitEnvironment()
    {
        try
        {
            System.exit(0);
        }
        catch (Exception e) { }
    }

    /// <summary>
    /// Shows the config as log
    /// </summary>
    private void ShowConfig()
    {
        String tmp =
                "---------------------Config---------------------\r\n" +
                        "| Mode: ";

        if (mode == 0)
        {
            tmp += "ServerClient\r\n" +
                    "| Internal Port: " +
                    Connection2.Port + "\r\n" +
                    "| Destination: " +
                    Connection1.Host + ":" +
                    Connection1.Port + "\r\n";

        }
        else if (mode == 1)
        {
            tmp += "ClientClient\r\n" +
                    "| Destination-Server: " +
                    Connection1.Host + ":" +
                    Connection1.Port + "\r\n" +
                    "| PortTunnel Server: " +
                    Connection2.Host + ":" +
                    Connection2.Port + "\r\n" +
                    "| PortTunnel Control-Port: " +
                    ControlPort + "\r\n";
        }
        else if (mode == 2)
        {
            tmp += "ServerServer\r\n" +
                    "| Internal Server-Port: " +
                    Connection1.Port + "\r\n" +
                    "| PortTunnel Server-Port: " +
                    Connection2.Port + "\r\n" +
                    "| PortTunnel Control-Port: " +
                    ControlPort + "\r\n";
        }

        tmp += "| HTTP: " + httpProtocol + "\r\n";

        tmp += "| SSL: " + destIsSsl + "\r\n";

        tmp += "| Timeout: " + Timeout + "\r\n";

        tmp += "------------------------------------------------\r\n";

        Log.Write(LogType.Info, "\r\n" + tmp);
    }

    /// <summary>
    /// Requests help
    /// </summary>
    /// <returns>Text of help file</returns>
    public static String GetHelp(){
        InputStream stream = PortTunnel.class.getResourceAsStream("/HelpFile.txt");
        BufferedInputStream buffer = new BufferedInputStream(stream);

        try {
            StringBuilder tmp = new StringBuilder();
            int tmpC;
            do{
                tmpC = buffer.read();
                if(tmpC == -1)
                    break;
                else
                    tmp.append((char)tmpC);
            }while(true);
            buffer.close();
            stream.close();
            return tmp.toString();
        }
        catch (Exception e) {
            return "";
        }
    }

    //-----------------------------Java specific-------------------------------

    public void AddListener(LogListener listener){
        LogListeners.add(listener);
    }

    public void AddListener(StreamLogListener listener){
        StreamLogListeners.add(listener);
    }

    public void setDestIsSsl(boolean isSsl){
        destIsSsl = isSsl;
    }

    public void setHttpProtocol(boolean usesHttp){
        httpProtocol = usesHttp;
    }

    public void setTicksBetweenBytes(int ticks){
        if(ticks >= 0)
            TicksBetweenBytes = ticks;
        else
            TicksBetweenBytes = 0;
    }

    public int getActiveConnections(){
        return ActiveConnections;
    }
}

File: PortTunnelLib/CSharp/PortTunnel.cs
/*
MIT License
Copyright (c) 2019 Patrick72762
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;

namespace PortTunnelLib
{
    /// <summary>
    /// PortTunnel main class
    /// </summary>
    public class PortTunnel
    {
        //-------------------------Private Attributes--------------------------
        private int mode;

        //---------------------------Inernal fields----------------------------
        internal Connection Connection1 { get; private set; }
        internal Connection Connection2 { get; private set; }
        internal int ControlPort { get; private set; }
        internal int Timeout { get; private set; }

        //----------------------------Public fields----------------------------
        private bool destIsSsl = false;
        private bool httpProtocol = false;
        private long ticksBetweenBytes = 10;
        /// <summary>
        /// Encrypted connection to destination server if set to true
        /// </summary>
        public bool DestIsSsl
        {
            get
            {
                return destIsSsl;
            }
            set
            {
                if (Status == 0)
                    destIsSsl = value;
            }
        }
        /// <summary>
        /// Replaces HTTP header if set to true
        /// </summary>
        public bool HttpProtocol
        {
            get
            {
                return httpProtocol;
            }
            set
            {
                if (Status == 0)
                    httpProtocol = value;
            }
        }
        /// <summary>
        /// Ticks between bytes in a tunnel
        /// </summary>
        public long TicksBetweenBytes
        {
            get
            {
                return ticksBetweenBytes;
            }
            set
            {
                if (value >= 0)
                    ticksBetweenBytes = value;
                else
                    ticksBetweenBytes = 0;
            }
        }

        /// <summary>
        /// Status of PortTunnel
        /// 0 - Stopped
        /// 1 - Active
        /// </summary>
        public int Status { get; private set; }
        /// <summary>
        /// Active tunnels of this PortTunnel
        /// </summary>
        public int ActiveConnections { get; internal set; } = 0;
        /// <summary>
        /// Consume to output or display log data
        /// </summary>
        public EventHandler<string> NewLogOutput;
        /// <summary>
        /// Consume to get stream data after stream is closed
        /// </summary>
        public EventHandler<string> NewStreamLogCompleted;

        //---------------------------Public Methods----------------------------
        /// <summary>
        /// Constructor for command line args
        /// </summary>
        /// <param name="args">Arguments for PortTunnel</param>
        public PortTunnel(string[] args)
        {
            Log.main = this;
            bool configured = true;

            string tmp = null;
            Dictionary<string, string> pairs = new Dictionary<string, string>();
            pairs.Add("-m", null);
            pairs.Add("-h1", null);
            pairs.Add("-p1", null);
            pairs.Add("-h2", null);
            pairs.Add("-p2", null);
            pairs.Add("-cp", null);
            pairs.Add("-t", null);

            NewLogOutput += (sender, text) =>
             {
                 Console.Write(text);
             };

            if (args.Length > 0)
            {
                if (args.Length > 1)
                {

                    foreach (string arg in args)
                    {
                        if (arg == "-http")
                        {
                            httpProtocol = true;
                        }
                        else if (arg == "-ssl")
                        {
                            destIsSsl = true;
                        }
                        else if (arg == "-log")
                        {
                            NewStreamLogCompleted += (sender, text) =>
                            {
                                string filename = ((string)text).Split('\r')[0]
                                    .Replace(' ', '-').Replace(":", "-Port-") + ".txt";

                                if (!Directory.Exists("Log"))
                                    Directory.CreateDirectory("Log");

                                File.WriteAllText("Log/" + filename, (string)text);
                            };
                        }
                        else
                        {
                            if (tmp == null)
                                tmp = arg;
                            else
                            {
                                try
                                {
                                    pairs[tmp] = arg;
                                }
                                catch
                                {
                                    Log.Write(Enums.LogType.Warning, "Unknown argument: " + tmp);
                                }
                                tmp = null;
                            }
                        }
                    }

                    if (pairs["-m"].ToLower() == "sc")
                    {
                        if (pairs["-h1"] != null && pairs["-p1"] != null && pairs["-p2"] != null)
                        {
                            try
                            {
                                Connection1 = new Connection(pairs["-h1"], int.Parse(pairs["-p1"]));
                                Connection2 = new Connection(int.Parse(pairs["-p2"]));
                            }
                            catch
                            {
                                Log.Write(Enums.LogType.Error, "Hostname or port not supported!");
                                configured = false;
                            }
                        }
                        else
                        {
                            Log.Write(Enums.LogType.Error, "Arguments not set for mode " + pairs["-m"]);
                            configured = false;
                        }
                    }
                    else if (pairs["-m"].ToLower() == "cc")
                    {
                        if (pairs["-h1"] != null && pairs["-p1"] != null &&
                            pairs["-h1"] != null && pairs["-p2"] != null && pairs["-cp"] != null)
                        {
                            try
                            {
                                Connection1 = new Connection(pairs["-h1"], int.Parse(pairs["-p1"]));
                                Connection2 = new Connection(pairs["-h2"], int.Parse(pairs["-p2"]));
                                ControlPort = int.Parse(pairs["-cp"]);
                            }
                            catch
                            {
                                Log.Write(Enums.LogType.Error, "Hostname or port not supported!");
                                configured = false;
                            }
                        }
                        else
                        {
                            Log.Write(Enums.LogType.Error, "Arguments not set for mode " + pairs["-m"]);
                            configured = false;
                        }
                    }
                    else if (pairs["-m"].ToLower() == "ss")
                    {
                        if (pairs["-p1"] != null && pairs["-p2"] != null && pairs["-cp"] != null)
                        {
                            try
                            {
                                Connection1 = new Connection(int.Parse(pairs["-p1"]));
                                Connection2 = new Connection(int.Parse(pairs["-p2"]));
                                ControlPort = int.Parse(pairs["-cp"]);
                            }
                            catch
                            {
                                Log.Write(Enums.LogType.Error, "Hostname or port not supported!");
                                configured = false;
                            }
                        }
                        else
                        {
                            Log.Write(Enums.LogType.Error, "Arguments not set for mode " + pairs["-m"]);
                            configured = false;
                        }
                    }
                    else
                    {
                        Log.Write(Enums.LogType.Error, "Unknown mode!");
                        configured = false;
                    }
                }
                else
                {
                    configured = false;
                    if(args[0] == "-h")
                    {
                        Log.Write(Enums.LogType.Info, "\r\n" + GetHelp());
                    }
                }
            }
            else
            {
                Log.Write(Enums.LogType.Error, "No args found!");
                configured = false;
            }

            try
            {
                Timeout = int.Parse(pairs["-t"]);
            }
            catch
            {
                Timeout = 3000;
            }

            if (!configured)
            {
                Log.Write(Enums.LogType.Info, "Exiting now");
                ExitEnvironment();
            }
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="connection1">Object</param>
        /// <param name="connection2">Object</param>
        /// <param name="timeout">Timeout for each tunnel</param>
        /// <param name="controlPort">PortTunnel communication port</param>
        public PortTunnel(Connection connection1, Connection connection2,
            int timeout, int controlPort = 0)
        {
            Log.main = this;

            Connection1 = connection1;
            Connection2 = connection2;
            ControlPort = controlPort;
            Timeout = timeout;
        }

        /// <summary>
        /// Starts PortTunnel
        /// </summary>
        /// <returns>false if config is not supported</returns>
        public bool Start()
        {
            bool supported = true;
            Status = 1;

            if (Connection1.Type == Enums.ConnectionType.Client &&
                Connection2.Type == Enums.ConnectionType.Server)
            {
                mode = 0;
                ServerClient Starter = new ServerClient(this);
                ShowConfig();
                Starter.Run();
            }
            else if (Connection1.Type == Enums.ConnectionType.Client &&
                Connection2.Type == Enums.ConnectionType.Client)
            {
                mode = 1;
                ClientClient Starter = new ClientClient(this);
                ShowConfig();
                Starter.Run();
            }
            else if (Connection1.Type == Enums.ConnectionType.Server &&
                Connection2.Type == Enums.ConnectionType.Server)
            {
                mode = 2;
                ServerServer Starter = new ServerServer(this);
                ShowConfig();
                Starter.Run();
            }
            else
            {
                supported = false;
                Log.Write(Enums.LogType.Error, "Config not supported");
            }

            return supported;
        }

        /// <summary>
        /// Stops PortTunnel
        /// </summary>
        public void Stop()
        {
            Status = 0;
        }

        /// <summary>
        /// Exits the main process
        /// </summary>
        public static void ExitEnvironment()
        {
            try
            {
                Environment.Exit(0);
            }
            catch { }
        }

        /// <summary>
        /// Shows the config as log
        /// </summary>
        public void ShowConfig()
        {
            string tmp =
                "---------------------Config---------------------\r\n" +
                "| Mode: ";

            if (mode == 0)
            {
                tmp += "ServerClient\r\n" +
                    "| Internal Port: " +
                    Connection2.Port + "\r\n" +
                    "| Destination: " +
                    Connection1.Host + ":" +
                    Connection1.Port + "\r\n";

            }
            else if (mode == 1)
            {
                tmp += "ClientClient\r\n" +
                    "| Destination-Server: " +
                    Connection1.Host + ":" +
                    Connection1.Port + "\r\n" +
                    "| PortTunnel Server: " +
                    Connection2.Host + ":" +
                    Connection2.Port + "\r\n" +
                    "| PortTunnel Control-Port: " +
                    ControlPort + "\r\n";
            }
            else if (mode == 2)
            {
                tmp += "ServerServer\r\n" +
                    "| Internal Server-Port: " +
                    Connection1.Port + "\r\n" +
                    "| PortTunnel Server-Port: " +
                    Connection2.Port + "\r\n" +
                    "| PortTunnel Control-Port: " +
                    ControlPort + "\r\n";
            }

            tmp += "| HTTP: " + HttpProtocol + "\r\n";

            tmp += "| SSL: " + DestIsSsl + "\r\n";

            tmp += "| Timeout: " + Timeout + "\r\n";

            tmp += "------------------------------------------------\r\n";

            Log.Write(Enums.LogType.Info, "\r\n" + tmp);
        }

        /// <summary>
        /// Requests help
        /// </summary>
        /// <returns>Text of help file</returns>
        public static string GetHelp()
        {
            Assembly _Assembly = Assembly.GetExecutingAssembly();
            Stream str = _Assembly
                .GetManifestResourceStream("PortTunnelLib.HelpFile.txt");
            StreamReader rd = new StreamReader(str);

            string tmp = rd.ReadToEnd();

            rd.Close();

            return tmp;
        }
    }
}
